File: Solutions/2958- Length of Longest Subarray With at Most K Frequency/Solution.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        map_count = defaultdict(int)
        left = 0
        right = 0
        answer = 0

        while left <= right and right < len(nums):
            if map_count[nums[right]] < k:
                map_count[nums[right]] += 1
                right += 1
            else:
                answer = max(answer, right - left)
                while map_count[nums[right]] >= k:
                    map_count[nums[left]] -= 1
                    left += 1

        return max(answer, right - left)
        

File: Solutions/2958- Length of Longest Subarray With at Most K Frequency/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int maxSubarrayLength(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        int left = 0;
        int right = 0;
        int answer = 0;

        while (left <= right && right < nums.length) {
            if (!map.containsKey(nums[right])) {
                map.put(nums[right], 1);
                right++;
            } else if (map.get(nums[right]) < k) {
                map.put(nums[right], map.get(nums[right]) + 1);
                right++;
            } else {
                answer = Math.max(answer, (right - left));
                while (map.get(nums[right]) >= k) {
                    map.put(nums[left], map.get(nums[left]) - 1);
                    left++;
                }
            }
        }

        return Math.max(answer, (right - left));
    }
}